models/repo_watch.go

```diff
@@ -74,66 +74,143 @@ func (repo *Repository) GetWatchers(page int) ([]*User, error) {
-	if err := act.Repo.getOwner(e); err != nil {
-		return fmt.Errorf("can't get repo owner: %v", err)
-	}
+func notifyWatchers(e Engine, actions ...*Action) error {
+	var watchers []*Watch
+	var repo *Repository
+	var err error
+	var permCode []bool
+	var permIssue []bool
+	var permPR []bool
+
+	for _, act := range actions {
+		repoChanged := repo == nil || repo.ID != act.RepoID
+
+		if repoChanged {
+			// Add feeds for user self and all watchers.
+			watchers, err = getWatchers(e, act.RepoID)
+			if err != nil {
+				return fmt.Errorf("get watchers: %v", err)
+			}
+		}
 
-	// Add feed for organization
-	if act.Repo.Owner.IsOrganization() && act.ActUserID != act.Repo.Owner.ID {
-		act.ID = 0
-		act.UserID = act.Repo.Owner.ID
+		// Add feed for actioner.
+		act.UserID = act.ActUserID
 		if _, err = e.InsertOne(act); err != nil {
 			return fmt.Errorf("insert new actioner: %v", err)
 		}
-	}
 
-	for i := range watches {
-		if act.ActUserID == watches[i].UserID {
-			continue
+		if repoChanged {
+			act.loadRepo()
+			repo = act.Repo
+
+			// check repo owner exist.
+			if err := act.Repo.getOwner(e); err != nil {
+				return fmt.Errorf("can't get repo owner: %v", err)
+			}
+		} else if act.Repo == nil {
+			act.Repo = repo
+		}
+
+		// Add feed for organization
+		if act.Repo.Owner.IsOrganization() && act.ActUserID != act.Repo.Owner.ID {
+			act.ID = 0
+			act.UserID = act.Repo.Owner.ID
+			if _, err = e.InsertOne(act); err != nil {
+				return fmt.Errorf("insert new actioner: %v", err)
+			}
 		}
 
-		act.ID = 0
-		act.UserID = watches[i].UserID
-		act.Repo.Units = nil
+		if len(actions) == 1 {
+			// More efficient to just check the code and not cache
+			for _, watcher := range watchers {
+				if act.ActUserID == watcher.UserID {
+					continue
+				}
+
+				act.ID = 0
+				act.UserID = watcher.UserID
+				act.Repo.Units = nil
+
+				switch act.OpType {
+				case ActionCommitRepo, ActionPushTag, ActionDeleteTag, ActionDeleteBranch:
+					if !act.Repo.checkUnitUser(e, act.UserID, false, UnitTypeCode) {
+						continue
+					}
+				case ActionCreateIssue, ActionCommentIssue, ActionCloseIssue, ActionReopenIssue:
+					if !act.Repo.checkUnitUser(e, act.UserID, false, UnitTypeIssues) {
+						continue
+					}
+				case ActionCreatePullRequest, ActionMergePullRequest, ActionClosePullRequest, ActionReopenPullRequest:
+					if !act.Repo.checkUnitUser(e, act.UserID, false, UnitTypePullRequests) {
+						continue
+					}
+				}
+
+				if _, err = e.InsertOne(act); err != nil {
+					return fmt.Errorf("insert new action: %v", err)
+				}
+			}
+			return nil
+		}
 
-		switch act.OpType {
-		case ActionCommitRepo, ActionPushTag, ActionDeleteTag, ActionDeleteBranch:
-			if !act.Repo.checkUnitUser(e, act.UserID, false, UnitTypeCode) {
-				continue
+		if repoChanged {
+			permCode = make([]bool, len(watchers))
+			permIssue = make([]bool, len(watchers))
+			permPR = make([]bool, len(watchers))
+			for i, watcher := range watchers {
```

Comment:
If I understand correctly, the last time it goes through this part `repoChanged` might be false, but there will still be some tasks pending.

Reply:
Er... I don't understand what you mean.

Reply:
This block of code is simply caching the results of the permissions checks - there's no tasks to perform at this point.

line 148 iterates across the watchers and does the tasks - that's not in the `if repoChanged` block.